Some Minebot improvements and additions

benbot16 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->
This PR does a bunch of different things to minebots that should hopefully make them more useful and more fun to play as.
Firstly, their active equipment has been changed:
- Minebots now have an integrated mining scanner
- Minebots can now mount plasma cutters and mining drills in addition to their standard PKA. They don't start with a plasma cutter, but they do start with a drill.
- The minebot drill, scanner, and plasma cutter can be replaced by using a new object of that type on the minebot (such as using a diamond drill to give the bot a better drill, or an advanced scanner to give it more scanning range).
- I also added several new upgrades for minebots, namely a module that weatherproofs them, an automatic ore scoop for player-controlled minebots, and a module that lets them carry a medipen to help downed miners. Sentient minebots can also take armor upgrades now, instead of having their armor arbitrarily reduced when being made sentient.

Additionally, I made some QOL changes to minebot examining and interactions. Players can now alt-click on minebots to have them drop their stored ore, instead of using a scanner on them. I also improved the examining text for minebots, since the old text was weirdly formatted and I needed to add more information. Sentient minebots now have a damage icon, as well was additional information in their status pane based on what modules they have equipped. I also added a few other interactions, such as EMP effects for minebots and the ability to repair a disabled bot.

Finally, I added a mining mode to minebots that allows them to mine while non-sentient. This probably needs a lot of looking-over, since I've never really done this sort of thing before, but it should allow minebots to mine and collect their mined ores fairly efficiently.

Meson vision's still a bit janky, but I'm not too sure how to fix that.

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->
Minebots right now are just kinda bad, and making them more useful should hopefully encourage miners to use them more.

This is probably imbalanced in some way, so any feedback in that regard would be helpful.

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>

<summary>Screenshots&Videos</summary>
Installing upgrades:

https://user-images.githubusercontent.com/9423435/177423265-261e5227-008b-4f17-af5c-2db6306f7d99.mp4

Automatic mining:

https://user-images.githubusercontent.com/9423435/177423330-07a6781c-c0d9-49d4-b7e4-bc566e1789c8.mp4

Player-controlled mining:

https://user-images.githubusercontent.com/9423435/177423382-2af06af5-641b-4b72-a78d-251395cb90b2.mp4

Repairing/Restarting:

https://user-images.githubusercontent.com/9423435/177423086-7f8a4a53-d58a-434e-8f81-9b3462b1b890.mp4

</details>

## Changelog
:cl:
refactor: Refactors minebot code. 
add: Minebots can now autonomously mine when set to mining mode. However, they won't avoid obstacles such as lava while this is active, so you should still keep an eye on them.
add: Minebots aren't gibbed on death, and can be revived with a welder.
tweak: Removing ore from a minebot is now done by alt-clicking the minebot.
add: Adds a weatherproofing module, an automatic ore pickup module, and a medipen module for minebots.
balance: The minebot AI upgrade doesn't nerf a minebot's stats anymore, however some upgrades have been changed to compensate for this.
balance: Minebots can now equip additional equipment such as plasma cutters, advanced drills, and mining scanners. These can be attached to a minebot by using them on said minebot.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

benbot16 committed:
Initial code

benbot16 committed:
Initial code

benbot16 committed:
Minebot AI stuff, comments, and wording

benbot16 committed:
Minebot AI stuff, comments, and wording

benbot16 committed:
Auto stash before merge of "master" and "BeeStation/master"
pain and suffering


Initial code


Minebot AI stuff, comments, and wording


More tweaks/fixes

benbot16 committed:
More tweaks/fixes

benbot16 committed:
More tweaks/fixes

benbot16 committed:
Auto stash before merge of "master" and "BeeStation/master"
pain and suffering


Initial code


Minebot AI stuff, comments, and wording


More tweaks/fixes

benbot16 committed:
Conflict fixes 2

benbot16 committed:
desc tweak

benbot16 committed:
desc tweak

benbot16 committed:
Fixes/tweaks to prior code + bugfixes

benbot16 committed:
Fixes/tweaks to prior code + bugfixes

benbot16 committed:
Removes a redundant check

benbot16 committed:
Removes a redundant check

benbot16 committed:
Comments + a bunch of readability tweaks

Is there such a thing as too many comments?

benbot16 committed:
Comments + a bunch of readability tweaks

Is there such a thing as too many comments?

benbot16 committed:
A few improvements on oldcode

- Removes damage prob
- Adjusts the examine text
- Makes EMPs deal damage
- Improves the attackby() chain for the minebot itself

benbot16 committed:
A few improvements on oldcode

- Removes damage prob
- Adjusts the examine text
- Makes EMPs deal damage
- Improves the attackby() chain for the minebot itself

benbot16 committed:
A couple other tweaks

benbot16 committed:
A couple other tweaks

benbot16 committed:
Fixes a few bugs and updates some code

- Changes some descs/comments
- Fixes KAs and cutters having spread when used by a minebot
- Changes some stuff to better adapt to coding conventions
- Makes the meson vision of the bot look less shit

benbot16 committed:
Fixes a few bugs and updates some code

- Changes some descs/comments
- Fixes KAs and cutters having spread when used by a minebot
- Changes some stuff to better adapt to coding conventions
- Makes the meson vision of the bot look less shit

benbot16 committed:
Wandering off bad

This should fix the dodge issues I had earlier

benbot16 committed:
Removes a death drop and simplifies target listing

benbot16 committed:
Minebot cooldown decrease change

benbot16 committed:
Desc update

benbot16 committed:
Conflict Fix

Fix for the mining sentience cooldown PR

benbot16 committed:
Conflict fix 2

benbot16 committed:
Conflict fix 3

Apparently I can't read.

PowerfulBacon commented on code/modules/mining/minebot.dm:
```diff
@@ -1,313 +1,699 @@
+		if(stat == DEAD)
+			adjustBruteLoss(-25)
+			to_chat(user, "<span class='info'>You repair and restart [src].</span>")
+			revive()
```
This should only revive if they have enough health to not instantly die

benbot16 replied:
Changed it to function that way.

PowerfulBacon commented on code/modules/mining/minebot.dm:
```diff
@@ -1,313 +1,699 @@
+	var/mode = MODE_MINING /// What mode the minebot is in
+	var/mining_enabled = FALSE /// Whether or not the minebot will mine new ores while in mining mode.
+	var/list/installed_upgrades /// A list of all the minebot's installed upgrades
+	var/obj/item/gun/energy/kinetic_accelerator/minebot/stored_pka /// The minebot's stored PKA
+	var/obj/item/gun/energy/plasmacutter/stored_cutter /// The minebot's stored plasma cutter
+	var/obj/item/pickaxe/drill/stored_drill /// The minebot's stored drill
+	var/obj/item/t_scanner/adv_mining_scanner/stored_scanner /// The minebot's stored mining scanner
```
These strong references can result in harddels

benbot16 replied:
I added some signals to clear them if they get deleted, let me know if I missed anything.

benbot16 committed:
A few adjustments to minebot equipment code

- Protects the minebot from contents explosions
- Adds signals to keep track of the minebot's equipment, and null it if it gets qdeled
- Tweaks repairing the minebot a bit
- EMP stun uses the SECONDS define

PowerfulBacon reviewed (approved):
Seems alright
